crates/server/src/services/core.rs

```diff
@@ -176,41 +176,37 @@ struct Inner<Digest: SupportedDigest> {
 impl<Digest: SupportedDigest> Inner<Digest> {
     // Load state from DataStore or initialize empty state, returning any
     // entries that are not yet part of a checkpoint.
-    async fn initialize(&mut self) -> Result<Vec<LogLeaf>, CoreServiceError> {
-        let initial = self.store.get_initial_leaves().await?.peekable();
-        pin_mut!(initial);
-
-        // If there are no stored log entries, initialize a new state
-        if initial.as_mut().peek().await.is_none() {
-            self.initialize_new().await?;
-            return Ok(vec![]);
-        }
+    async fn initialize(&mut self) -> Result<(), CoreServiceError> {
+        tracing::debug!("Initializing CoreService");
 
-        // Reconstruct internal state from previously-stored log entires
-        let state = self.state.get_mut();
-        let mut pending_entries = vec![];
-        let mut last_checkpoint = Hash::<Digest>::of(()).into();
+        let published = self.store.get_all_validated_records().await?.peekable();
+        pin_mut!(published);
 
-        while let Some(res) = initial.next().await {
-            let InitialLeaf { leaf, checkpoint } = res?;
+        // If there are no published records, initialize a new state
+        if published.as_mut().peek().await.is_none() {
+            tracing::debug!("No existing records; initializing new state");
+            return self.initialize_new().await;
+        }
 
-            // Push the entry itself
-            state.push_entry(leaf.clone());
+        // Reconstruct internal state from previously-stored data
+        let mut checkpoints = self.store.get_all_checkpoints().await?;
+        let mut checkpoints_by_len: HashMap<usize, MapCheckpoint> = Default::default();
+        while let Some(checkpoint) = checkpoints.next().await {
+            let checkpoint = checkpoint?;
+            checkpoints_by_len.insert(checkpoint.log_length as usize, checkpoint);
+        }
 
-            // Update per-checkpoint state and track any entries that aren't
-            // yet part of a checkpoint
-            if let Some(checkpoint) = checkpoint {
-                if last_checkpoint != checkpoint {
-                    state.checkpoint(); // for the side-effect of updating map_index
-                    last_checkpoint = checkpoint;
-                }
-                pending_entries.clear();
-            } else {
-                pending_entries.push(leaf);
+        let state = self.state.get_mut();
+        while let Some(entry) = published.next().await {
+            state.push_entry(entry?);
+            if let Some(stored_checkpoint) = checkpoints_by_len.get(&state.log.length()) {
+                // Validate stored checkpoint (and update internal state as a side-effect)
+                let computed_checkpoint = state.checkpoint();
+                assert!(stored_checkpoint == &computed_checkpoint);
```

Comment:
@peterhuene I believe this avoids the checkpoint issue we discussed, and additionally validates the stored checkpoints.